File: 2405.1/pythonchk/security/PublicDataAssignedToPrivateArrayChecker/PublicDataAssignedToPrivateArray.py
import copy


class BadTest2:

    def __init__(self):
        self.__colors = None
        __colors = []

    def get_colors(self):
        return self.__colors[:]

    def set_colors(self, colors: [str]):
        self.__colors = colors  # @violation


class BadTest:
    __colors = None

    def get_colors(self):
        return self.__colors[:]

    def set_colors(self, colors):
        self.__colors = colors  # @violation


color = ['red', 'orange', 'black', 'white', 'blue']

bad_test = BadTest()
bad_test.set_colors(color)
color[0] = 'none'
print(bad_test.get_colors())  # ['none', 'orange', 'black', 'white', 'blue']


class GoodTest:
    __colors = None

    def get_colors(self):
        return self.__colors[:]

    def set_colors1(self, colors):
        self.__colors = colors[:]  # good

    def set_colors2(self, colors):
        self.__colors = copy.deepcopy(colors)  # good


color = ['red', 'orange', 'black', 'white', 'blue']

good_test = GoodTest()
good_test.set_colors1(color)  # ['red', 'orange', 'black', 'white', 'blue']
color[0] = 'none'
print(good_test.get_colors())  # ['red', 'orange', 'black', 'white', 'blue']


class BadUserObj:
    __private_variable = []

    def __init__(self):
        pass

    # private 배열에 외부 값을 바로 대입하는 public 메소드를 사용하는 경우 취약함
    def set_private_member(self, num):
        self.__private_variable = num  # @violation


class GoodUserObj:
    def __init__(self):
        self.__privateVariable = []

    # private 배열에 외부 값을 바로 대입하는 경우 [:]를 사용하여
    # 외부와 내부의 배열이 서로 참조되지 않도록 해야 함
    def set_private_member(self, num):
        self.__privateVariable = num[:]  # good

File: 2405.1/kotlinchk/security/PublicDataAssignedToPrivateArrayChecker/PublicDataAssignedToPrivateArray.kt
package testcase.rule.security.PublicDataAssignedToPrivateArrayChecker

object TestCase {
	class Violation1 {
		private lateinit var colors : Array<String?>

		public fun getColors(): Array<String?> {
			return this.colors
		}
		public fun setColors(colors: Array<String?>) {
			this.colors = colors							// @violation
		}
	}
	fun violationTest1() {
		val data: Array<String?> = arrayOf("red", "orange", "black", "white", "blue")
		val color = Violation1()
		color.setColors(data)
		println(color.getColors().contentToString())		// [red, orange, black, white, blue]
		data[0] = "none"
		println(color.getColors().contentToString())		// [red, orange, black, white, blue]
	}

	class Violation2 {
		private lateinit var colors : MutableList<String?>

		fun getColors(): MutableList<String?> {
			return this.colors
		}
		fun setColors(colors1: MutableList<String?>) {
			this.colors = colors1							// @violation
		}
	}
	fun violationTest2() {
		val data: MutableList<String?> = mutableListOf("red", "orange", "black", "white", "blue")
		val color = Violation2()
		color.setColors(data)
		println(color.getColors())							// [red, orange, black, white, blue]
		data[0] = "none"
		println(color.getColors())							// [red, orange, black, white, blue]
	}

	class Violation3 {
		private var colors = mutableListOf<String?>()

		fun getColors(): MutableList<String?> {
			return this.colors
		}
		fun setColors(colors1: MutableList<String?>) {
			colors = colors1								// @violation
		}
	}
	fun violationTest3() {
		val data: MutableList<String?> = mutableListOf("red", "orange", "black", "white", "blue")
		val color = Violation3()
		color.setColors(data)
		println(color.getColors())							// [red, orange, black, white, blue]
		data[0] = "none"
		println(color.getColors())							// [red, orange, black, white, blue]
	}

	class Good1 {
		private var colors : Array<String?> = arrayOf()

		fun getColors(): Array<String?> {
			return this.colors
		}
		fun setColors(color: Array<String?>) {
			this.colors = color.clone()						// good
		}
	}
	fun goodTest() {
		val data: Array<String?> = arrayOf("red", "orange", "black", "white", "blue")
		val color = Good1()
		color.setColors(data)
		println(color.getColors().contentToString())		// [red, orange, black, white, blue]
		data[0] = "none"
		println(color.getColors().contentToString())		// [red, orange, black, white, blue]
	}

	class Good2 {
		private var colors = mutableListOf<String?>()

		fun getColors(): MutableList<String?> {
			return this.colors
		}
		fun setColors(color: MutableList<String?>) {
			this.colors = color.toMutableList()				// good
		}
	}
	fun goodTest2() {
		val data: MutableList<String?> = mutableListOf("red", "orange", "black", "white", "blue")
		val color = Good2()
		color.setColors(data)
		println(color.getColors())							// [red, orange, black, white, blue]
		data[0] = "none"
		println(color.getColors())							// [red, orange, black, white, blue]
	}

	class Good3 {
		private lateinit var colors : Array<String?>

		private fun getColors(): Array<String?> {
			return this.colors
		}
		private fun setColors(colors: Array<String?>) {
			this.colors = colors							// good
		}
	}

	class Good4 {
		private lateinit var colors : Array<String?>

		protected fun getColors(): Array<String?> {
			return this.colors
		}
		protected fun setColors(colors: Array<String?>) {
			this.colors = colors							// good
		}
	}
}